File: examples/contracts/src/test/java/contracts/purity/LedgerContract.java
package contracts.purity;

import example.purity.Ledger;
import org.bmc4j.BmcContractsFor;
import org.bmc4j.Ensures;
import org.bmc4j.Requires;

/**
 * A contract on an <b>impure</b> method, for the purity-audit demo. The {@code @Ensures} here is
 * actually <em>true</em> of the return value ({@code Ledger.record} does return a value
 * {@code >= the amount} for non-negative inputs), so a naive enforce-proof would go GREEN —
 * which is precisely the trap: the contract says nothing about, and the replace-stub silently
 * drops, the method's mutation of {@code Ledger.total}.
 *
 * <p>The <b>purity audit</b> rejects this contract before any proof can reuse it: certifying
 * {@code Ledger.record} reaches a {@code PUTSTATIC} of {@code Ledger.total} (a heap write to
 * pre-existing state), so it fails with a {@code ContractPurityError} naming that instruction.
 * That rejection is what {@code proofs.purity.PurityAuditDemoTest} asserts.
 *
 * <p>The generated {@code LedgerContract__BmcEnforce} enforce-proof is therefore <em>expected to be
 * rejected</em>, not run as an ordinary proof — the module's {@code test} task excludes it (see
 * {@code build.gradle.kts}); the hand-written demo test drives the rejection deterministically.
 */
@BmcContractsFor(Ledger.class)
interface LedgerContract {

    @Requires("nonNegative")
    @Ensures("atLeastAmount")
    int record(int amount);

    static boolean nonNegative(int amount) {
        return amount >= 0 && amount <= 1_000;
    }

    static boolean atLeastAmount(int result, int amount) {
        // True of the RETURN value alone — yet the method is impure, which the audit catches.
        return result >= amount;
    }
}

File: examples/contracts-kotlin/src/test/kotlin/contracts/purity/LedgerContract.kt
package contracts.purity

import example.purity.Ledger
import org.bmc4j.BmcContractsFor
import org.bmc4j.Ensures
import org.bmc4j.Requires

/**
 * A contract on the impure [Ledger.record], in the standard **object-host** shape (a plain `object`
 * with ordinary member `fun` predicates). Its `@Ensures` happens to be true (the new total is at least
 * the old, for non-negative input), so the enforce-proof alone would PASS — which is exactly the false
 * green the purity audit refuses to allow. Because `record` mutates the receiver (`this.total`), the
 * audit rejects this contract at proof time with a `ContractPurityError` naming the `PUTFIELD` — an
 * UNCONDITIONAL build failure (no `@ExpectEnforce` can bless an impure target).
 *
 * The generated `enforce__record` is therefore EXCLUDED from this module's `test` task (see
 * `build.gradle.kts`); `proofs.purity.PurityAuditDemoTest` documents the rejection deterministically.
 * Removing that exclusion is itself the regression check — the build then goes red with the audit's
 * message.
 */
@BmcContractsFor(Ledger::class)
object LedgerContract {

    @Requires("nonNeg")
    @Ensures("grew")
    fun record(amount: Int): Int = error("mirror")

    // Plain member predicates — no companion, no @JvmStatic.
    fun nonNeg(self: Ledger, amount: Int): Boolean = amount in 0..1000
    fun grew(result: Int, self: Ledger, amount: Int): Boolean = result >= self.total
}
